homeassistant/components/tuya/fan.py

```
@@ -61,68 +61,103 @@ def async_discover_device(device_ids: list[str]) -> None:
+        )
+        if int_type := self.find_dpcode(
+            dpcodes, dptype=DPType.INTEGER, prefer_function=True
         ):
-            if DPCode.FAN_SPEED_ENUM in self.device.function:
-                self.dp_code_speed_enum = DPCode.FAN_SPEED_ENUM
-            else:
-                self.dp_code_speed_enum = DPCode.SPEED
+            self._attr_supported_features |= SUPPORT_SET_SPEED
+            self._speed = int_type
+        elif enum_type := self.find_dpcode(
+            dpcodes, dptype=DPType.ENUM, prefer_function=True
+        ):
+            self._attr_supported_features |= SUPPORT_SET_SPEED
+            self._speeds = enum_type
 
-            data = json.loads(self.device.function[self.dp_code_speed_enum].values).get(
-                "range"
-            )
-            if data:
-                self.air_purifier_speed_range_len = len(data)
-                self.air_purifier_speed_range_enum = data
+        if dpcode := self.find_dpcode(
+            (DPCode.SWITCH_HORIZONTAL, DPCode.SWITCH_VERTICAL), prefer_function=True
+        ):
+            self._oscillate = dpcode
+            self._attr_supported_features |= SUPPORT_OSCILLATE
+
+        if enum_type := self.find_dpcode(
+            DPCode.FAN_DIRECTION, dptype=DPType.ENUM, prefer_function=True
+        ):
+            self._direction = enum_type
+            self._attr_supported_features |= SUPPORT_DIRECTION
 
     def set_preset_mode(self, preset_mode: str) -> None:
         """Set the preset mode of the fan."""
-        self._send_command([{"code": DPCode.MODE, "value": preset_mode}])
+        if self._presets is None:
+            return
+        self._send_command([{"code": self._presets.dpcode, "value": preset_mode}])
 
     def set_direction(self, direction: str) -> None:
         """Set the direction of the fan."""
-        self._send_command([{"code": DPCode.FAN_DIRECTION, "value": direction}])
+        if self._direction is None:
+            return
```

Should this be silently swallowed or error raised because it's not supported ?

---

In reality, this will never be called, as the supported feature protects that. This is merely resolving typing.